rfcs/20240325-reorganize-gpu/README.md

```diff
@@ -0,0 +1,115 @@
+abstractions and therefore have dependencies on OpenCL and nGEN even though
+there is no need in them. Furthermore, oneDNN now has generic SYCL
+kernels that can be used on a variety of different architectures
+that are supported by the SYCL ecosystem. The SYCL kernels also use
+abstractions that are built on top of the basic GPU abstractions and
+therefore have the same issue.
+
+This RFC proposes a new organization of GPU kernels and abstractions
+to clearly separate independent functionality, get rid of unnecessary
+dependencies and have a flexible enough architecture to make adding support
+for new vendors easier.
+
+## Proposal
+
+Eeorganize th
+
+Reorganize the GPU kernels and abstractions according the the following
+schema: Vendor / Technology
+
+With this schema the GPU directory will have subdirectories that
+correspond to the vendors: `intel`, `nvidia`, `amd`, `generic`, etc.
+Each of the subdirectories will have technology specific sub-subdirectories:
+`sycl`, `ocl`, `jit`, etc.
+
+Pros:
+* The schema provides enough flexibility to enable new vendor and extend
+the supported ones
+* Clustering functionality and abstractions around vendors is very
+convenient as a vendor can share the same functionality/abstractions across
+different technologies (e.g. the compute layer for Intel vendor)
+* Currently, functionality and abstractions are fully or partially
+clustered around vendors therefore the new schema should not cause a lot of
+confusion among the developers, it should also positively affect the
+implementation cost
+* The schema also provides sufficient configurability, e.g
+the generic vendor can be enabled along with nvidia or amd, or it can be
+enabled individually
+
+```bash
+├── cpu/ # CPU code
+├── sycl/ # A common place for basic CPU and GPU SYCL abstractions, e.g. `sycl_cpu_engine_t`, `sycl_gpu_engine_t`, etc.
+└── gpu/ # GPU code. Basic GPU code resides in the GPU directory directly, e.g. `gpu_engine_t`.
+    ├── intel/ # Everything that is related to Intel
+    │   ├── compute/ # Compute layer abstractions
+    │   ├── ocl/ # OpenCL kernels and abstractions (very similar to the current state of `gpu/ocl` directory)
+    │   ├── jit/ # JIT kernels/generators and abstractions (very similar to the current state of `gpu/jit` directory)
+    │   ├── sycl/ # Intel related SYCL functionality/abstractions (e.g. SYCL kernels that use eSIMD, Intel specific extensions, etc)
+    │   └── ...
+    ├── nvidia/ # Everything that is related to NVIDIA
+    │   ├── sycl # NVIDIA related functionality/abstractions (cuDNN/cuBLAS based and CUDA specific SYCL kernels)
+    │   └── ...
+    ├── amd/ # Everything that is related to AMD
+    │   ├── sycl/ # AMD related functionality/abstractions (MIOpen/rocBLAS based and HIP specific SYCL kernels)
+    │   └── ...
+    └── generic/ # Everything that is related to generic kernels
+        └── sycl/ # generic SYCL kernels and related abstractions
```

Comment:
What would be the criteria for an OpenCL/SYCL kernel to not be generic? Most of our OpenCL kernels are generic in theory, so long as the runtime supports the necessary features. Do we have a method for separating dispatching/configuration heuristics from implementation?

Reply:
The SYCL specs defines what's generic and what's not: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:generic-vs-non-generic. 

I would say that a SYCL code that uses any vendor specific extensions would be considered as non-generic (e.g. [sycl_ext_intel_grf_size](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_intel_grf_size.asciidoc), etc).

The main driver for this proposal is our collaboration with UXL community. We agreed to provide generic SYCL kernels that should work with any vendor if they are supported by the compiler. In addition, we need to pave the way for more vendors to be supported. Therefore, we need to look beyond Intel vendor when reviewing the proposed architecture.